src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -573,22 +412,40 @@ private async ValueTask WriteAsyncChunked<TIOAdapter>(TIOAdapter writeAdapter, R
         private ValueTask WriteSingleChunk<TIOAdapter>(TIOAdapter writeAdapter, ReadOnlyMemory<byte> buffer)
             where TIOAdapter : struct, ISslIOAdapter
         {
-            // Request a write IO slot.
-            Task ioSlot = writeAdapter.WriteLockAsync();
-            if (!ioSlot.IsCompletedSuccessfully)
-            {
-                // Operation is async and has been queued, return.
-                return WaitForWriteIOSlot(writeAdapter, ioSlot, buffer);
-            }
-
             byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length + FrameOverhead);
             byte[] outBuffer = rentedBuffer;
 
-            SecurityStatusPal status = EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+            SecurityStatusPal status;
+            int encryptedBytes;
+            while (true)
+            {
+                status = EncryptData(buffer, ref outBuffer, out encryptedBytes);
+
+                // This should be rare, when renegotiation happens exactly when we want to write.
+                if (status.ErrorCode == SecurityStatusPalErrorCode.TryAgain)
+                {
+                    TaskCompletionSource<bool>? waiter = _handshakeWaiter;
+                    if (waiter != null)
+                    {
+                        Task waiterTask  = writeAdapter.WaitAsync(waiter);
+                        // We finished synchronously waiting for renegotiation. Try it again.
+                        if (waiterTask.IsCompletedSuccessfully)
+                        {
+                            continue;
+                        }
+
+                        return WaitAndWriteAsync(waiterTask, rentedBuffer);
+                    }
+
+                    continue;
+                    // We are on Async path now. We need to wait for the Lock as well as for the write when it is finished.
+                }
+
+                break;
```

Comment:
Nit: to avoid indenting all the code above and to make the common case clearer, consider inverting the if condition above such that the break is moved up above and is there only part indented

Reply:
updated